services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/crt/S3CrtRequestBodyStreamAdapter.java

```diff
@@ -16,323 +16,33 @@
-        // we encounter either of the terminal events, COMPLETE or ERROR.
-        while (true) {
-            // The supplier API requires that we fill the buffer entirely.
-            if (!outBuffer.hasRemaining()) {
-                break;
-            }
-
-            if (eventBuffer.isEmpty() && pending == 0) {
-                pending = DEFAULT_REQUEST_SIZE;
-                subscription.request(pending);
-            }
-
-            Event ev = takeFirstEvent();
-
-            // Discard the event if it's not for the current subscriber
-            if (!ev.subscriber().equals(subscriber)) {
-                LOG.debug(() -> "Received an event for a previous publisher. Discarding. Event was: " + ev);
-                continue;
-            }
-
-            switch (ev.type()) {
-                case DATA:
-                    ByteBuffer srcBuffer = ((DataEvent) ev).data();
-
-                    ByteBuffer bufferToWrite = srcBuffer.duplicate();
-                    int nBytesToWrite = Math.min(outBuffer.remaining(), srcBuffer.remaining());
-
-                    // src is larger, create a resized view to prevent
-                    // buffer overflow in the subsequent put() call
-                    if (bufferToWrite.remaining() > nBytesToWrite) {
-                        bufferToWrite.limit(bufferToWrite.position() + nBytesToWrite);
-                    }
-
-                    outBuffer.put(bufferToWrite);
-                    srcBuffer.position(bufferToWrite.limit());
-
-                    if (!srcBuffer.hasRemaining()) {
-                        --pending;
-                    } else {
-                        eventBuffer.push(ev);
-                    }
-
-                    break;
-
-                case COMPLETE:
-                    // Leave this event in the queue so that if getRequestData
-                    // gets call after the stream is already done, we pop it off again.
-                    eventBuffer.push(ev);
-                    pending = 0;
-                    return true;
-
-                case ERROR:
-                    // Leave this event in the queue so that if getRequestData
-                    // gets call after the stream is already done, we pop it off again.
-                    eventBuffer.push(ev);
-                    Throwable t = ((ErrorEvent) ev).error();
-                    if (t instanceof RuntimeException) {
-                        throw (RuntimeException) t;
-                    }
-                    throw new RuntimeException(t);
-
-                default:
-                    // In case new event types are introduced that this loop doesn't account for
-                    throw new IllegalStateException("Unknown event type: " + ev.type());
-            }
-        }
-
-        return false;
+        return requestBodySubscriber.blockingTransferTo(outBuffer) == ByteBufferStoringSubscriber.TransferResult.END_OF_STREAM;
```

Comment:
What thread is blocking here?

Reply:
As discussed offline, CRT thread is blocking here; it requires the byte buffer to be completely filled. Added a comment

Reply:
I don't like that. Seems to scale poorly...

Reply:
+1 I don't like it either. This will for sure cause issues with larger part size. We are having conversations with CRT team about this